finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+        new_network::{
+            connection_manager::{Authentication, SessionHandler},
+            DataCommand, Multiaddr, Protocol,
+        },
+        NodeIndex, SessionId,
+    };
+    use aleph_primitives::{AuthorityId, KEY_TYPE};
+    use codec::Encode;
+    use sc_network::{
+        multiaddr::Protocol as ScProtocol, Multiaddr as ScMultiaddr, PeerId as ScPeerId,
+    };
+    use sp_keystore::{testing::KeyStore, CryptoStore};
+    use std::{collections::HashSet, iter, net::Ipv4Addr, sync::Arc, thread::sleep, time};
+
+    const NUM_NODES: u8 = 7;
+    const MS_COOLDOWN: u64 = 200;
+
+    fn addresses() -> Vec<Multiaddr> {
+        (0..NUM_NODES)
+            .map(|id| {
+                ScMultiaddr::empty()
+                    .with(ScProtocol::Ip4(Ipv4Addr::new(192, 168, 1, id)))
+                    .with(ScProtocol::Tcp(30333))
+                    .with(ScProtocol::P2p(ScPeerId::random().into()))
+            })
+            .collect()
+    }
+
+    async fn keyboxes() -> Vec<KeyBox> {
+        let num_keyboxes: usize = NUM_NODES.into();
+        let keystore = Arc::new(KeyStore::new());
+        let mut auth_ids = Vec::with_capacity(num_keyboxes);
+        for _ in 0..num_keyboxes {
+            let pk = keystore.ed25519_generate_new(KEY_TYPE, None).await.unwrap();
+            auth_ids.push(AuthorityId::from(pk));
+        }
+        let mut result = Vec::with_capacity(num_keyboxes);
+        for i in 0..num_keyboxes {
+            result.push(KeyBox::new(
+                NodeIndex(i),
+                AuthorityVerifier::new(auth_ids.clone()),
+                AuthorityPen::new(auth_ids[i].clone(), keystore.clone())
+                    .await
+                    .expect("The keys should sign successfully"),
+            ));
+        }
+        result
+    }
+
+    async fn build() -> (Discovery, Vec<SessionHandler>) {
+        let mut handlers = Vec::new();
+        for (keybox, address) in keyboxes().await.into_iter().zip(addresses()) {
+            handlers.push(
+                SessionHandler::new(keybox, SessionId(43), vec![address.into()])
+                    .await
+                    .unwrap(),
+            );
+        }
+        (
+            Discovery::new(time::Duration::from_millis(MS_COOLDOWN)),
+            handlers,
+        )
+    }
+
+    #[tokio::test]
+    async fn broadcasts_when_clueless() {
+        let (mut discovery, mut handlers) = build().await;
+        let handler = &mut handlers[0];
+        let mut messages = discovery.discover_authorities(handler);
+        assert_eq!(messages.len(), 1);
+        let message = messages.pop().unwrap();
+        assert_eq!(message.1, DataCommand::Broadcast);
+        match message.0 {
+            DiscoveryMessage::AuthenticationBroadcast(authentication) => {
+                assert_eq!(authentication.encode(), handler.authentication().encode())
+            }
+            _ => panic!("Expected an authentication broadcast, got {:?}", message.0),
+        }
```

Comment:
Egh, I would prefer not to do that, equality of authentications is kinda icky?

Reply:
I meant exactly the check you do here, i.e. pattern-matching + `==` on authentication (this is what `assert_eq` does I believe)

Reply:
Rather than doing `==` on authentications I compare their encodings. It's a workaround, but maybe it's stupid, I'll check how annoying it is to derive `PartialEq` for them.

Reply:
It wasn't, I guess I should just always derive that. <_<"

Anyway, done.